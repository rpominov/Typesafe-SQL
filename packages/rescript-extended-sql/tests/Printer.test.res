open Jest

// (Js.Array2.t<string>, paramLinks<int>)

let showStr = x => x->Js.Json.string->Js.Json.stringify

let rec showParams = (indent, params: Printer.paramLinks<int>) =>
  "{\n  " ++
  indent ++
  params
  ->Js.Array2.map(({name, link}) =>
    name ++
    ": " ++
    switch link {
    | Plain(n) => `$${n->Js.Int.toString}`
    | Raw(options) =>
      `Raw(\n    ${indent}${options
        ->Js.Array2.map(showStr)
        ->Js.Array2.joinWith("\n    " ++ indent)}\n${indent}  )`
    | Batch(separator, params') => `Batch("${separator}" ${showParams(indent ++ "  ", params')})`
    }
  )
  ->Js.Array2.joinWith("\n  " ++ indent) ++
  "\n" ++
  indent ++ "}"

let expectToMatchSnapshot = makeSnapshotMatcher((
  val: (Js.Array2.t<string>, Printer.paramLinks<int>),
) => {
  let (codes, params) = val
  "SQL = [\n  " ++
  codes->Js.Array2.map(showStr)->Js.Array2.joinWith("\n  ") ++
  "\n]\nPARAMS = " ++
  showParams("", params)
})

each(
  [
    "SELECT 1",
    "-- inline\nSELECT /* block */1",
    "SELECT :foo = :bar",
    "SELECT :num:raw<1|2>",
    "SELECT :num:raw<1,|> 2",
    "SELECT :num:raw<<<'<|>'|||'2'>>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<(:foo, :bar)>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:batch<:bar>)>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:raw<foo|bar>)>>",
    "INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo /* <comment> */)>>",
  ],
  "%s",
  code => (code->Parser.parse->getOkExn(__LOC__)).ast->Printer.print->expectToMatchSnapshot,
)
