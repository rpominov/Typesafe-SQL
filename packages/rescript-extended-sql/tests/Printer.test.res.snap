// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`-- inline
SELECT /* block */1 1`] = `
SQL = [
  "\\nSELECT 1"
]
PARAMS = {
  
}
`;

exports[`INSERT INTO test (foo, bar) VALUES :values:batch<(:foo, :bar)> 1`] = `
SQL = [
  "INSERT INTO test (foo, bar) VALUES ($1, $2),($3, $4)"
]
PARAMS = {
  values: Batch("," {
    foo: $1
    bar: $2
  })
}
`;

exports[`INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo /* <comment> */)>> 1`] = `
SQL = [
  "INSERT INTO test (foo, bar) VALUES ($1 ),($2 )"
]
PARAMS = {
  values: Batch("," {
    foo: $1
  })
}
`;

exports[`INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:batch<:bar>)>> 1`] = `
SQL = [
  "INSERT INTO test (foo, bar) VALUES ($1,$2),($3,$4)"
]
PARAMS = {
  values: Batch("," {
    foo: Batch("," {
      bar: $1
    })
  })
}
`;

exports[`INSERT INTO test (foo, bar) VALUES :values:batch<<(:foo:raw<foo|bar>)>> 1`] = `
SQL = [
  "INSERT INTO test (foo, bar) VALUES (foo),(foo)"
  "INSERT INTO test (foo, bar) VALUES (foo),(bar)"
  "INSERT INTO test (foo, bar) VALUES (bar),(foo)"
  "INSERT INTO test (foo, bar) VALUES (bar),(bar)"
]
PARAMS = {
  values: Batch("," {
    foo: Raw(
      "foo"
      "bar"
    )
  })
}
`;

exports[`SELECT :foo = :bar 1`] = `
SQL = [
  "SELECT $1 = $2"
]
PARAMS = {
  foo: $1
  bar: $2
}
`;

exports[`SELECT :num:raw<<<'<|>'|||'2'>>> 1`] = `
SQL = [
  "SELECT '<|>'"
  "SELECT '2'"
]
PARAMS = {
  num: Raw(
    "'<|>'"
    "'2'"
  )
}
`;

exports[`SELECT :num:raw<1,|> 2 1`] = `
SQL = [
  "SELECT 1, 2"
  "SELECT  2"
]
PARAMS = {
  num: Raw(
    "1,"
    ""
  )
}
`;

exports[`SELECT :num:raw<1|2> 1`] = `
SQL = [
  "SELECT 1"
  "SELECT 2"
]
PARAMS = {
  num: Raw(
    "1"
    "2"
  )
}
`;

exports[`SELECT 1 1`] = `
SQL = [
  "SELECT 1"
]
PARAMS = {
  
}
`;
